// eslint-plugin-custom-rules/src/id-camel.js
/**
 * @fileoverview Enforce id prop is camel cased.
 */

// ----------------------------------------------------------------------------
// Rule Definition
// ----------------------------------------------------------------------------

import { elementType, getProp, getPropValue } from 'jsx-ast-utils'
import { generateObjSchema } from './schema'
import { camelCase } from 'lodash'

const errorMessage = (id) => `id ${id} must be camel cased`

const schema = generateObjSchema()

module.exports = {
  meta: {
    docs: {},
    schema: [schema],
  },

  create: context => ({
    JSXOpeningElement: (node) => {
      const type = elementType(node)

      const id = getPropValue(getProp(node.attributes, 'id'))

      if (!id) {
        return
      }

      // split string with space
      const idArr = id.split(' ')

      const hasFailed = idArr.some(id => id !== camelCase(id))

      if (!hasFailed) {
        return
      }

      let failedId = ''
      if (hasFailed) {
        failedId = idArr.filter(id => id !== camelCase(id)).join(' ')
      }

      context.report({
        node,
        message: errorMessage(failedId),
      })
    }
  })
}
