// eslint-plugin-custom-rules/lib/id-camel.js
'use strict';

var _jsxAstUtils = require('jsx-ast-utils');

var _schema = require('./schema');

var _lodash = require('lodash');

var errorMessage = function errorMessage(id) {
  return 'id ' + id + ' must be camel cased';
}; /**
    * @fileoverview Enforce id prop is camel cased.
    */

// ----------------------------------------------------------------------------
// Rule Definition
// ----------------------------------------------------------------------------

var schema = (0, _schema.generateObjSchema)();

module.exports = {
  meta: {
    docs: {},
    schema: [schema]
  },

  create: function create(context) {
    return {
      JSXOpeningElement: function JSXOpeningElement(node) {
        var type = (0, _jsxAstUtils.elementType)(node);

        var id = (0, _jsxAstUtils.getPropValue)((0, _jsxAstUtils.getProp)(node.attributes, 'id'));

        if (!id) {
          return;
        }

        // split string with space
        var idArr = id.split(' ');

        var hasFailed = idArr.some(function (id) {
          return id !== (0, _lodash.camelCase)(id);
        });

        if (!hasFailed) {
          return;
        }

        var failedId = '';
        if (hasFailed) {
          failedId = idArr.filter(function (id) {
            return id !== (0, _lodash.camelCase)(id);
          }).join(' ');
        }

        context.report({
          node: node,
          message: errorMessage(failedId)
        });
      }
    };
  }
};
